linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,322 @@
+                {
+                    result->thread_count = 0;
+                    result->task_count = 0;
+                    result->quitting = 0;
+                    (void)interlocked_exchange(&result->pool_state, POOL_STATE_UNINIT);
+
+                    result->task_list = NULL;
+                    result->task_list_last = NULL;
+                    goto all_ok;
+                }
+                free(result->thread_handle_list);
+            }
+            free(result);
+            result = NULL;
+        }
+    }
+all_ok:
+    return result;
+}
+
+void threadpool_destroy(THREADPOOL_HANDLE threadpool)
+{
+    if (threadpool == NULL)
+    {
+        // do nothing
+        LogError("Invalid arguments: THREADPOOL_HANDLE threadpool: %p", threadpool);
+    }
+    else
+    {
+        interlocked_exchange(&threadpool->quitting, 1);
+
+        for (size_t index = 0; index < threadpool->list_index; index++)
+        {
+            int dont_care;
+            ThreadAPI_Join(threadpool->thread_handle_list[index], &dont_care);
+        }
+
+        free(threadpool->thread_handle_list);
+
+        srw_lock_destroy(threadpool->srw_lock);
+        free(threadpool);
+    }
+}
+
+int threadpool_schedule_work(THREADPOOL_HANDLE threadpool, THREADPOOL_WORK_FUNCTION work_function, void* work_function_context)
+{
+    int result;
+    if (threadpool == NULL || work_function == NULL)
+    {
+        LogError("Invalid arguments: THREADPOOL_HANDLE threadpool: %p, THREADPOOL_TASK_FUNC work_function: %p", threadpool, work_function);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        THREADPOOL_TASK* task_item = malloc(sizeof(THREADPOOL_TASK));
+        if (task_item == NULL)
+        {
+            LogError("Failure allocating THREADPOOL_TASK structure");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            task_item->next = NULL;
+            task_item->task_param = work_function_context;
+            task_item->task_func = work_function;
+
+            // Need to add a lock here
+            srw_lock_acquire_exclusive(threadpool->srw_lock);
+
+            if (threadpool->task_list == NULL)
+            {
+                threadpool->task_list = task_item;
+            }
+            else
+            {
+                threadpool->task_list_last->next = task_item;
+            }
+            threadpool->task_list_last = task_item;
+            (void)interlocked_increment(&threadpool->task_count);
+
+            srw_lock_release_exclusive(threadpool->srw_lock);
+
+            // unlock
+
+            if (interlocked_add(&threadpool->pool_state, 0) == POOL_STATE_IDLE)
+            {
+                // Idle signal that we have work
+                wake_by_address_single(&threadpool->task_count);
+                result = 0;
+                goto all_ok;
+            }
+            else
+            {
+                // Check to see
+                if (threadpool->thread_count < threadpool->max_thread_count)
```

Comment:
>threadpool->thread_count  [](http://example.com/codeflow?start=20&length=25)

unsafe check, needs some kind of interlocked_compare_exchange or the InterlockedHL thing that can do conditional compare-exchange operations

Reply:
so probably interlocked_add, then interlocked_compare_exchange to increment and only create a new thread if that succeeds (and loop as appropriate).

Reply:
You can't do interlocked_add, test, and modify as separate steps. That's why we need compare-exchange in a loop

Right now, two threads can decide the count is below max and both add a new thread even though it exceeds max.

In other words, it's broken for example when the count is MAX-1 and 2 threads execute the check around the same time.